### PR TITLE
GT-960 Tips tutorial frequency

### PR DESCRIPTION
**Pull request by frett**
Show the tips tutorial 3 times per tool

### CODE REVIEW

File: library/base/src/main/java/org/cru/godtools/base/Settings.kt [hunk 1]
```diff
@@ -51,7 +51,7 @@ class Settings @Inject internal constructor(
         const val FEATURE_TUTORIAL_ONBOARDING = "tutorialOnboarding"
         const val FEATURE_TUTORIAL_TRAINING = "tutorialTraining"
         const val FEATURE_TUTORIAL_LIVE_SHARE = "tutorialLiveShare."
-        const val FEATURE_TUTORIAL_TIPS = "tutorialTips"
+        const val FEATURE_TUTORIAL_TIPS = "tutorialTips."
```
Comment (TheNoodleMoose) [inline library/base/src/main/java/org/cru/godtools/base/Settings.kt:54]: Why did you add a period here?

File: app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt [hunk 2]
```diff
@@ -133,7 +133,8 @@ class ToolDetailsFragment() : BasePlatformFragment<ToolDetailsFragmentBinding>(R
 
         activity?.run {
             openToolActivity(code, tool.type, locale, showTips = true)
-            if (!settings.isFeatureDiscovered(FEATURE_TUTORIAL_TIPS)) {
+            if (settings.getFeatureDiscoveredCount("$FEATURE_TUTORIAL_TIPS$code") < 3) {
+                settings.setFeatureDiscovered("$FEATURE_TUTORIAL_TIPS$code")
```
Comment (TheNoodleMoose) [inline app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt:137]: I assume that ```settings.getFeatureDiscoveredCount``` returns an Int value for how many times the tutorial tips have been shown for a specific tool. And we decide which tool to check by passing ```$code``` in the function as an argument. Where are those values stored?

File: ui/tutorial-renderer/src/main/java/org/cru/godtools/tutorial/activity/TutorialActivity.kt [hunk 3]
```diff
@@ -79,7 +79,7 @@ class TutorialActivity : BaseActivity<TutorialActivityBinding>(), TutorialCallba
     }
 
     override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
-        R.id.onboarding_action_skip -> {
+        R.id.onboarding_action_skip, R.id.action_tips_skip -> {
```
Comment (TheNoodleMoose) [inline ui/tutorial-renderer/src/main/java/org/cru/godtools/tutorial/activity/TutorialActivity.kt:82]: I assume skip logic is already implemented, and this is actually just letting that logic actually run once the skip menu item is pressed for tutorial tips. So is it safe to assume if you didn't add this, and I tried pressing skip that it would do nothing?

File: ui/tutorial-renderer/src/main/res/menu/tutorial_tips_menu.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <item
+        android:id="@+id/action_tips_skip"
+        android:title="@string/tutorial_tips_action_skip"
+        app:showAsAction="always" />
```
Comment (TheNoodleMoose) [inline ui/tutorial-renderer/src/main/res/menu/tutorial_tips_menu.xml:8]: Does this attribute change styling in some way?

**Review by TheNoodleMoose — APPROVED** [review]

File: app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt (hunk 2 above)
Comment (frett) [reply app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt:137]: yes, the `Settings` object has all the feature discovery logic. using Flipper you can look at the underlying SharedPreferences that actually stores the raw data.
Screen Shot 2020-10-08 at 11 37 03 AM

Comment (frett) [reply app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt:137]: `$code` is the tool code (e.g. `kgp`, `fourlaws`)

File: library/base/src/main/java/org/cru/godtools/base/Settings.kt (hunk 1 above)
Comment (frett) [reply library/base/src/main/java/org/cru/godtools/base/Settings.kt:54]: because I'm concatenating the tool code to the end of the key. so the key will look like: `tutorialTips.kgp`

File: ui/tutorial-renderer/src/main/java/org/cru/godtools/tutorial/activity/TutorialActivity.kt (hunk 3 above)
Comment (frett) [reply ui/tutorial-renderer/src/main/java/org/cru/godtools/tutorial/activity/TutorialActivity.kt:82]: yeah, I opted to provide unique ids to the menu items for different tutorials that way if the logic ever needs to diverge I'm ready to handle the unique case for tips vs onboarding

File: ui/tutorial-renderer/src/main/res/menu/tutorial_tips_menu.xml (hunk 4 above)
Comment (frett) [reply ui/tutorial-renderer/src/main/res/menu/tutorial_tips_menu.xml:8]: so, this is menu XML, it controls what menus items are present in the toolbar. by default the toolbar decides how many items to show directly as actions, and the rest get put into the action overflow menu. this attribute forces this action to always be shown directly and never be put in the action overflow menu

Comment (TheNoodleMoose) [reply ui/tutorial-renderer/src/main/res/menu/tutorial_tips_menu.xml:8]: Ah okay makes sense!

File: library/base/src/main/java/org/cru/godtools/base/Settings.kt (hunk 1 above)
Comment (TheNoodleMoose) [reply library/base/src/main/java/org/cru/godtools/base/Settings.kt:54]: Oh, gotcha!

File: app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt (hunk 2 above)
Comment (TheNoodleMoose) [reply app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt:137]: I was able to get Flipper connected and find the stored raw data, thanks!
